File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    int dp[1001][1001];
    
    bool helper(int i,int j,string&s){
        
         if(i>j)
            return false;
        
        if(i==j)
            return dp[i][j] = true;
        if(j == i+1)
            return dp[i][j] = s[i]==s[j];
        
        if(dp[i][j]!=-1)
            return dp[i][j];
        
        helper(i+1,j,s);
        helper(i,j-1,s);
        
        if(s[i]==s[j])
            dp[i][j] = true & helper(i+1,j-1,s);
        
        else dp[i][j] = 0;
        
        return dp[i][j];
    }
    
    string longestPalindrome(string s) {
        memset(dp,-1,sizeof dp);
        int n = s.size();
        helper(0,n-1,s);
        
        int ans = 1;
        string curr = "";
        curr+=s[0];
        
        for(int i=0;i<n;i++){
            for(int j=i+1;j<n;j++){
                if(dp[i][j] and j-i+1>ans){
                    curr = s.substr(i,j-i+1);
                    ans = j - i + 1;
                }
            }
        }
        
        return curr;
    }
};

File: 2025-and-before/0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    private int n;
    private int[][] dp, palinDp;

    // Function to check if the substring s[i:j+1] is a palindrome
    private int isPalin(int i, int j, String s) {
        if (i > j)
            return palinDp[i][j] = 1;  // An empty string is considered a palindrome
        if (i == j)
            return palinDp[i][j] = 1;  // A single character is a palindrome

        if (palinDp[i][j] != -1)
            return palinDp[i][j];

        if (s.charAt(i) != s.charAt(j)) {
            return palinDp[i][j] = 0;
        }

        return palinDp[i][j] = isPalin(i + 1, j - 1, s);  // Check the inner substring
    }

    // Helper function to compute the length of the longest palindromic substring
    private int helper(int i, int j, String s) {
        if (i > j)
            return 0;

        if (dp[i][j] != -1)
            return dp[i][j];

        int ans = 0;
        if (isPalin(i, j, s) == 1)
            ans = j - i + 1;

        helper(i + 1, j, s);
        helper(i, j - 1, s);

        return dp[i][j] = ans;
    }

    public String longestPalindrome(String s) {
        n = s.length();
        dp = new int[n][n];
        palinDp = new int[n][n];

        // Initialize dp and palinDp arrays
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1);
            Arrays.fill(palinDp[i], -1);
        }

        helper(0, n - 1, s);

        int maxLen = 1, start = 0, end = 0;

        // Find the maximum length of the palindromic substring
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (isPalin(i, j, s) == 1 && j - i + 1 > maxLen) {
                    start = i;
                    end = j;
                    maxLen = j - i + 1;
                }
            }
        }

        // Corrected substring method call
        return s.substring(start, end + 1);
    }
}
